### PR TITLE
PYIC-7556 refactor ipv middleware tests

### PR DESCRIPTION
**Pull request by thebauSoftwire**
### What changed
Separates the ipv middleware tests in their own files, tidies them up and removes duplicates

### Why did it change
Keeping all the tests in a single file was getting unwieldy and becoming difficult to debug tests because the requests were always being mutated. 

### Issue tracking

- [PYIC-7556](https://govukverify.atlassian.net/browse/PYIC-7556)

[PYIC-7556]: https://govukverify.atlassian.net/browse/PYIC-7556?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/app/ipv/tests/formHandleCoiDetailsCheck.test.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+  const middleware: typeof import("../middleware") = proxyquire(
+    "../middleware",
+    {
+      "../../services/coreBackService": coreBackServiceStub,
+    },
+  );
+
+  beforeEach(() => {
+    next = sinon.fake() as any;
```
Comment (thebauSoftwire) [inline src/app/ipv/tests/formHandleCoiDetailsCheck.test.ts]: I can't get `sinon.reset()` working, not sure what I'm missing :/ so for now have had to reassign the variables this way instead to reset any stubs/fakes

File: src/app/ipv/tests/checkFormRadioButtons.test.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    body: {},
+    params: { pageId: "page-ipv-identity-document-start" },
+    csrfToken: sinon.fake(),
+    session: {
+      currentPage: "page-ipv-identity-document-start",
+      ipvSessionId: "ipv-session-id",
+      save: sinon.fake.yields(null),
+    },
+    log: { error: sinon.fake() },
+  } as any;
```
Comment (thebauSoftwire) [inline src/app/ipv/tests/checkFormRadioButtons.test.ts:16]: I think it was already mentioned in another one of Joe's PRs, but it'd be nice if there was some builder/factory methods to create test data but this PR was getting quite big

File: src/app/ipv/tests/checkFormRadioButtons.test.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    locals: { contactUsUrl: "contactUrl", deleteAccountUrl: "deleteAccount" },
+  } as any;
+
+  const next = sinon.fake() as any;
+
+  beforeEach(() => {
+    res.render = sinon.fake();
+  });
+
+  it("should render if journey is not defined", async function () {
```
Comment (Joe-Edwards-GDS) [inline src/app/ipv/tests/checkFormRadioButtons.test.ts]: I'm going to try and contain myself on the test refactoring, since it's not really anything to do with you, but this is pretty bad...

Can we specify _what_ it is supposed to render? e.g. `should re-render the same page with an error if no option is selected`

And ideally also assert that `res.render` gets appropriate arguments (i.e. the same page as the input, and an error param)

File: src/app/ipv/tests/checkFormRadioButtons.test.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    const req = { ...testReq, body: { journey: undefined } };
+    await checkFormRadioButtonSelected(req, res, next);
+
+    expect(res.render).to.have.been.called;
+    expect(next).to.have.not.been.calledOnce;
+  });
+
+  it("should not render if journey is defined", async function () {
```
Comment (Joe-Edwards-GDS) [inline src/app/ipv/tests/checkFormRadioButtons.test.ts]: (and then here perhaps something like `should pass to next if an option is selected`)

File: src/app/ipv/tests/formHandleCoiDetailsCheck.test.ts (hunk 1 above)
Comment (Joe-Edwards-GDS) [reply src/app/ipv/tests/formHandleCoiDetailsCheck.test.ts]: Strange - I don't really know `sinon` well enough to explain it, I'd need to play with it to figure out what's going on :shrug:

(I'm a bit more familiar with Jest - but that has its own idiosyncracies)

File: src/app/ipv/tests/formHandleUpdateDetailsCheckBox.test.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    render: sinon.fake(),
+    log: { info: sinon.fake(), error: sinon.fake() },
+    locals: { contactUsUrl: "contactUrl", deleteAccountUrl: "deleteAccount" },
+  } as any;
+
+  beforeEach(() => {
+    next = sinon.fake() as any;
+  });
+
+  const updateDetailsCombosTestData = [
```
Comment (Joe-Edwards-GDS) [inline src/app/ipv/tests/formHandleUpdateDetailsCheckBox.test.ts]: Is this the same list as in the other test - can we share it?

(If it helps I wouldn't mind combining these two test files, since they're pretty closely related!)

File: src/app/ipv/tests/handleJourneyActionRequest.test.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    coreBackServiceStub.postJourneyEvent = callBack;
+    callBack.onCall(0).returns(clientEventResponse);
+
+    await middleware.handleJourneyActionRequest(testReq, res, next);
+    expect(next).to.have.been.calledWith(
+      sinon.match.has("message", "Client Response redirect url is missing"),
+    );
+  });
+
+  it("should call next with an error message given redirect url is missinf from CRI event response", async () => {
```
Comment (Joe-Edwards-GDS) [inline src/app/ipv/tests/handleJourneyActionRequest.test.ts]: Typo: `missinf`

Comment (Joe-Edwards-GDS) [reply src/app/ipv/tests/handleJourneyActionRequest.test.ts]: (FWIW I'm probably going to change these a little bit anyway...)

File: src/app/ipv/tests/handleJourneyActionRequest.test.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    coreBackServiceStub.postJourneyEvent = callBack;
+    callBack.onCall(0).returns(criEventResponse);
+
+    await middleware.handleJourneyActionRequest(testReq, res, next);
+    expect(next).to.have.been.calledWith(
+      sinon.match.has("message", "CRI response RedirectUrl is missing"),
+    );
+  });
+
+  ["ukPassport", "drivingLicence", "end", "attempt-recovery"].forEach(
```
Comment (Joe-Edwards-GDS) [inline src/app/ipv/tests/handleJourneyActionRequest.test.ts]: I don't think these are actually special any more(?) - we can probably just have a single test case with a test journey event (to make it clear that the event doesn't matter).

File: src/app/ipv/tests/handleJourneyActionRequest.test.ts [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        const req = { ...testReq, body: { journey } };
+        await middleware.handleJourneyActionRequest(req, res, next);
+        expect(
+          coreBackServiceStub.postJourneyEvent.firstCall,
+        ).to.have.been.calledWith(req, journey, req.session.currentPage);
+      });
+    },
+  );
+
+  it("should postJourneyEvent with build-client-oauth-response and use ip address from header when not present in session", async function () {
```
Comment (Joe-Edwards-GDS) [inline src/app/ipv/tests/handleJourneyActionRequest.test.ts]: Is this specific to `build-client-oauth-response`? Should it just be an assertion that the ip address is included on any journey action request?

File: src/app/ipv/tests/handleJourneyPageRequest.test.ts [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+  );
+
+  beforeEach(() => {
+    res.render = sinon.fake();
+    next = sinon.fake() as any;
+    coreBackServiceStub.getProvenIdentityUserDetails = sinon.spy();
+  });
+
+  context("handling page-ipv-reuse journey route", () => {
+    const nameTestCases = [
```
Comment (Joe-Edwards-GDS) [inline src/app/ipv/tests/handleJourneyPageRequest.test.ts]: It feels wrong to have all these cases here - I think we should cover the mapping edge cases in `reuseHelper.test.ts`, and keep this focused on the behaviour from the middleware.

File: src/app/ipv/tests/processAction.test.ts [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+      testReq,
+      "next",
+    );
+
+    expect(testReq.session.currentPageStatusCode).to.equal(418);
+    expect(res.redirect).to.have.been.calledWith(`/ipv/page/a-page-id`);
+  });
+
+  it("should be redirected to a valid redirectURL given a valid CRI event response", async function () {
+    const request =
```
Comment (Joe-Edwards-GDS) [inline src/app/ipv/tests/processAction.test.ts]: Does this need to be a real value?

**Review by Joe-Edwards-GDS — APPROVED** [review]
This is fine, but I've left a few suggestions to tidy up a bit further.

TBH the best thing would be to break up `middleware.ts` itself, but we can save that for later. The test breakdown gives some reasonable lines to try and split it up along!

**Review by Joe-Edwards-GDS — DISMISSED** [review]
Thanks!

**Review by shivanshuit914 — APPROVED** [review]
